### PR TITLE
Regional Partner support in Teachercon Registration form

### PR DESCRIPTION
**Pull request by Hamms**

### CODE REVIEW

**Comment by Hamms** [comment]
[TC registration form for partners - routing/authentication](https://trello.com/c/eATB0nGl/387-tc-registration-form-for-partners-routing-authentication)

File: apps/src/code-studio/pd/teachercon1819_registration/Teachercon1819Registration.jsx [hunk 1]
```diff
@@ -13,9 +13,10 @@ import { TeacherSeatAcceptanceOptions } from '@cdo/apps/generated/pd/teachercon1
 export default class Teachercon1819Registration extends FormController {
   static propTypes = {
     ...FormController.propTypes,
-    applicationId: PropTypes.number.isRequired,
+    applicationId: PropTypes.number,
+    regionalPartnerId: PropTypes.number,
     applicationType: PropTypes.string.isRequired,
-    course: PropTypes.string.isRequired,
+    course: PropTypes.string,
```
Comment (aoby) [inline apps/src/code-studio/pd/teachercon1819_registration/Teachercon1819Registration.jsx:17]: Is `course` no longer required?

File: dashboard/app/controllers/api/v1/pd/teachercon1819_registrations_controller.rb [hunk 2]
```diff
@@ -1,9 +1,41 @@
...
+  load_and_authorize_resource :application, class: 'Pd::Application::ApplicationBase', id_param: :applicationId, except: :create_partner
 
   def new_form
     ::Pd::Teachercon1819Registration.new(
       pd_application_id: @application.id
     )
   end
+
+  def create_partner
+    regional_partner_id = params.try(:[], :regionalPartnerId)
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/teachercon1819_registrations_controller.rb]: I don't think params can be `nil` in a controller action.

File: dashboard/app/controllers/api/v1/pd/teachercon1819_registrations_controller.rb [hunk 3]
```diff
@@ -1,9 +1,41 @@
...
+    regional_partner = RegionalPartner.find(regional_partner_id)
+    unless regional_partner && current_user && current_user.regional_partners.include?(regional_partner)
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/pd/teachercon1819_registrations_controller.rb]: Note, `find` will raise an error if the id doesn't exist, so `regional_partner` here can't be nil

File: dashboard/db/migrate/20180124234033_add_regional_partner_to_pd_teachercon1819_registration.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddRegionalPartnerToPdTeachercon1819Registration < ActiveRecord::Migration[5.0]
+  def change
+    add_reference :pd_teachercon1819_registrations, :regional_partner, foreign_key: true
+  end
+end
```
Comment (aoby) [inline dashboard/db/migrate/20180124234033_add_regional_partner_to_pd_teachercon1819_registration.rb]: For good measure, let's pull out the migration into its own PR.

File: apps/src/code-studio/pd/teachercon1819_registration/Teachercon1819Registration.jsx (hunk 1 above)
Comment (Hamms) [reply apps/src/code-studio/pd/teachercon1819_registration/Teachercon1819Registration.jsx:17]: it isn't; it's now only used by the teacher and facilitator versions

**Comment by Hamms** [comment]
The partner registration is not a totally different form, just a different version like the Teacher and Facilitator versions.

It does not require an application_guid because partners don't have applications; we instead get all the information we need from the associated RegionalPartner object

**Comment by Hamms** [comment]
PTAL
